_drafts/ble-throughput-primer.md

```diff
@@ -0,0 +1,297 @@
+
+{:.no_toc}
+
+#### Maximum Link Layer Data Payload Throughput
+
+With the three rules we mentioned about transmissions on the Bluetooth Radio in mind, let's take a look at the procedure to transmit a maximally sized LL packet.
+
+- Side A sends a maximum size LL data packet (**27** bytes of data in a **41** byte payload) which takes **328μs** (`41 bytes * 8 bits / byte * 1Mbps`) to transmit
+- Side B receieves the packet and waits `T_IFS` (**150μs**)
+- Side B ACKs the LL packet it received in **80μs** (**0** bytes of data)
+- Side A waits `T_IFS` before sending any more data
+
+Here's an example exchange of two packets of data in one _Connection Event_:
+![Data exchange](img/ble-throughput/connection_event.png)
+
+The time it takes to transmit one packet can be computed as:
+
+`328μs data packet + 150μs T_IFS + 80μs ACK + 150μs T_IFS = 708μs`
+
+During this time period, **27** bytes of actual data can be transmitted which takes **216μs**.
+
+This yields a raw data throughput of:
+
+`(216μs / 708μs) * 1Mbps = 305,084 bits/second = ~0.381 Mbps`
+
+### L2CAP Channels & Payloads
+
+L2CAP is the next layer above the Link Layer and is packed inside the **Data Payloads** of the Link Layer packets. L2CAP allows:
+
+- unrelated data flows to be multiplexed across different _logical channels_
+- fragmenting and de-fragmenting data across multiple LL Packets (since an L2CAP payload can span multiple LL Packets)
+
+There are only a few different L2CAP channels used for Bluetooth Low Energy:
+
+1. **Security Manager protocol** (**SMP**) - for BLE security setup
+2. **Attribute protocol** (**ATT**) - Android & iOS both offer APIs that allow 3rd party apps to control transfers over this channel.
+3. **LE L2CAP Connection Oriented Channels** (**CoC**) - Custom channels which are advantageous for streaming applications. Apple exposed APIs for L2CAP Channels in iOS 11[^2]. Android also exposed APIs as part of Android 10 [^3]. Since both of the mobile APIs are still fairly "new" and many phones aren't running OS updates which support these features, most devices which need to talk to a mobile device still need to use ATT. If your application does not need to send data to a mobile device, it's definitely worth trying to use CoCs!
+
+{:.no_toc}
+
+#### The L2CAP Packet
+
+Below is a diagram of the layout of an L2CAP packet. As you can see there are 4 bytes of overhead (Length + Channel ID) per packet sent.
+
+![Data exchange](img/ble-throughput/l2cap_packet.png)
+
+### Attribute Protocol (ATT) Packet
+
+In the _L2CAP Information Payload_ we have the ATT packet. This is the packet structure the _GATT_ protocol uses. For an excellent article about GATT itself, check out [this post](https://interrupt.memfault.com/blog/bluetooth-low-energy-a-primer#gatt-services-and-characteristics) on the topic!
```

Comment:
I'd honestly pitch the post a little more here. It's a killer post and deserves a bit more mention than "this post"

Maybe: For an excellent article about GATT itself, check out Mohammad Afaneh's thorough [BLE Primer post].